azkaban-common/src/main/java/azkaban/imagemgmt/permission/PermissionManagerImpl.java

```diff
@@ -0,0 +1,148 @@
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package azkaban.imagemgmt.permission;
+
+import azkaban.imagemgmt.daos.ImageTypeDao;
+import azkaban.imagemgmt.exeception.ImageMgmtException;
+import azkaban.imagemgmt.models.ImageOwnership;
+import azkaban.imagemgmt.models.ImageOwnership.Role;
+import azkaban.imagemgmt.models.ImageType;
+import azkaban.user.Permission;
+import azkaban.user.Permission.Type;
+import java.util.List;
+import java.util.Optional;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.apache.commons.collections4.CollectionUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class defines and manages the permission for accessing image management APIs. The
+ * permissions are defined on image type. The permissions are created using user role and access
+ * type. This class internally accesses image type cache to get image type metadata.
+ */
+@Singleton
+public class PermissionManagerImpl implements PermissionManager {
+
+  private static final Logger log = LoggerFactory.getLogger(PermissionManagerImpl.class);
+
+  private final ImageTypeDao imageTypeDao;
+
+  /**
+   * Permission for image type owner of role ADMIN. ADMIN owner has complete access to image
+   * management APIs including add/delete of new member as image type owner.
+   */
+  private static final Permission ADMIN_PERMISSION = new Permission(Type.ADMIN);
+  /**
+   * Permission for image type owner of role MEMBER. MEMBER owner has complete access to image
+   * management APIs except add/delete of new member as image type owner.
+   */
+  private static final Permission MEMBER_PERMISSION = new Permission(Type.CREATE, Type.GET,
+      Type.UPDATE, Type.DELETE);
+  /**
+   * The default permission is only GET access to the image management APIs.
+   */
+  private static final Permission READ_ONLY_PERMISSION = new Permission(Type.GET);
+
+  @Inject
+  public PermissionManagerImpl(final ImageTypeDao imageTypeDao) {
+    this.imageTypeDao = imageTypeDao;
+  }
+
+  /**
+   * Checks the permission based on image type name, user id and Permission type.
+   *
+   * @param imageTypeName
+   * @param userId
+   * @param type
+   * @return boolean
+   */
+  @Override
+  public boolean hasPermission(final String imageTypeName, final String userId, final Type type) {
+    // Gets the image type metadata including ownerships.
+    final Optional<ImageType> optionalImageType =
+        this.imageTypeDao.getImageTypeWithOwnershipsByName(imageTypeName);
+    boolean hasPermission = false;
+    if (optionalImageType.isPresent()) {
```

Comment:
If this `hasPermission` method is called as part of a new image creation request then `optionalImageType` will be empty (because it doesn't exist yet) and an exception will be thrown stopping the creation attempt with not enough permissions reason. How can a user create a new image type?

Reply:
Azkaban admin defined as part of azkaban-users.xml can create image types add the moment. Azkaban admin check is added as part of hasImageManagementPermission method in LoginAbstractAzkabanServlet.java. On the other hand, the ownership defined as part of image type has owners with roles such as ADMIN and MEMBER. So, initially only Azkaban admin (Azkaban dev or service account admin defined in azkaban-users.xml) can create a new image type with owners. Once the owners are added for an image type they can access the image management APIs. Owner with ADMIN role can access all the image management APIs including add/remove image type owners (This two api to be added in future). Owner with member role can access all the APIs except add/remove image type owners.